### ISSUE TITLE
Apply jsd_token_clip before gathering per token pointwise value together Result Negative JSD loss

### ISSUE THREAD

**Issue by HaowenGuan** [siyan-zhao/OPSD#2]
Thanks for sharing your source code. Your work is very interesting.
While trying your code, I get negative training loss, and I found the following issue:

In your JSD loss implementation, you apply the following token_clipping while the JSD loss values is of the shape: (B, T, V).

```
        # Per-token clipping: cap each token's divergence value
        if token_clip is not None:
            jsd = jsd.clamp(max=token_clip)
```

This end up clipping each datapoint loss value to below, for example, token_clip=0.06.
It is apparently wrong, as it turn to be "per-datapoint" clipping instead of  "per-token" clipping, and it results a negative value in final loss.
(datapoint loss value in kl_div loss can be negative, but summing them together should always be >= 0. When we clip big positive datapoint value to be <= 0.06, the sum end up being negative.)

I suppose you should reduce from (B, T, V) to (B, T) before clipping:

```
jsd = jsd.sum(dim=-1)
```

**Comment by siyan-zhao** [comment]
Hi, thank you for the careful read! You are right that our code is actually doing per-token per-vocab clipping before gathering and it can become negative value, which is more granular than per-token level clipping after gathering. We empirically find per-vocab clipping works better than after gathered. Thanks for pointing out the confusion and we will clarify more in the new version of the paper!

**Comment by HaowenGuan** [comment]
Thanks for your reply, that is an interesting finding. I look forward to the new revision of the paper.

**Comment by rteehas** [comment]
Hi, I also was experiencing negative loss (for this reason I believe). If you find that per-vocab clipping works better does that mean that negative loss is to be expected / not an issue? It seems odd

**Comment by siyan-zhao** [comment]
@rteehas hi yes negative loss is expected with this current version
